packages/pyright-internal/src/analyzer/checker.ts

```diff
@@ -2830,25 +2830,36 @@ export class Checker extends ParseTreeWalker {
             constraints
         );
 
-        const returnDiag = new DiagnosticAddendum();
-        if (
-            !isNever(overloadReturnType) &&
-            !this._evaluator.assignType(
-                implReturnType,
-                overloadReturnType,
-                returnDiag.createAddendum(),
-                constraints,
-                AssignTypeFlags.Default
-            )
-        ) {
-            returnDiag.addMessage(
-                LocAddendum.functionReturnTypeMismatch().format({
-                    sourceType: this._evaluator.printType(overloadReturnType),
-                    destType: this._evaluator.printType(implReturnType),
-                })
-            );
-            diag?.addAddendum(returnDiag);
-            isConsistent = false;
+        // based overload consistency part 1: The implementation return type must overlap with each overload
+        // when distributing unions (i.e., at least one member of the impl union is assignable to the overload).
+        if (!isNever(overloadReturnType)) {
+            const returnDiag = new DiagnosticAddendum();
+            let ok = false;
+            if (isUnion(implReturnType)) {
+                // Succeeds if any union member of the implementation overlaps with the overload return type.
+                for (const sub of implReturnType.priv.subtypes) {
+                    if (
+                        this._evaluator.assignType(overloadReturnType, sub) ||
+                        this._evaluator.assignType(sub, overloadReturnType)
+                    ) {
+                        ok = true;
+                        break;
+                    }
+                }
+            } else {
+                ok = this._evaluator.assignType(implReturnType, overloadReturnType);
+            }
```

Comment:
you can use `doForEachSubtype` to iterate over possible unions more easily

Reply:
i need to break, and do different behavior if it's not a union. wdyt?

Reply:
i need to break, and do different behavior if it's not a union. wdyt?